### DISCUSSION TITLE
issue: append issue with data is not iterable

### DISCUSSION THREAD

**Discussion by dayhaysoos** [react-hook-form/react-hook-form#7527] [Q&A] [answered]
### Version Number

7.11.0

### Codesandbox/Expo snack

https://codesandbox.io/s/silly-joana-u51tu?file=/src/ContributionInput.js

### Steps to reproduce

I actually have no idea how to reproduce, the Code Sandbox works as expected, however, when I try adding a field in my project I get an error when I run `append` that says "data is not iterable". 

### Expected behaviour

It's actually behaving as expected in the code sandbox, just getting an error in my actual project. Hoping someone would have some ideas 

### What browsers are you seeing the problem on?

Chrome

### Relevant log output

```shell
Uncaught TypeError: data is not iterable
    at append (index.esm.js?34a0:373)
    at batchStateUpdate (index.esm.js?34a0:484)
    at append$1 (index.esm.js?34a0:529)
    at onClick (contribution-input.js?a4f3:35)
    at HTMLUnknownElement.callCallback (react-dom.development.js?61bb:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js?61bb:3994)
    at invokeGuardedCallback (react-dom.development.js?61bb:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js?61bb:4070)
    at executeDispatch (react-dom.development.js?61bb:8243)
    at processDispatchQueueItemsInOrder (react-dom.development.js?61bb:8275)
    at processDispatchQueue (react-dom.development.js?61bb:8288)
    at dispatchEventsForPlugins (react-dom.development.js?61bb:8299)
    at eval (react-dom.development.js?61bb:8508)
    at batchedEventUpdates$1 (react-dom.development.js?61bb:22396)
    at batchedEventUpdates (react-dom.development.js?61bb:3745)
    at dispatchEventForPluginEventSystem (react-dom.development.js?61bb:8507)
    at attemptToDispatchEvent (react-dom.development.js?61bb:6005)
    at dispatchEvent (react-dom.development.js?61bb:5924)
    at unstable_runWithPriority (scheduler.development.js?3069:468)
    at runWithPriority$1 (react-dom.development.js?61bb:11276)
    at discreteUpdates$1 (react-dom.development.js?61bb:22413)
    at discreteUpdates (react-dom.development.js?61bb:3756)
    at dispatchDiscreteEvent (react-dom.development.js?61bb:5889)
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
hey, @dayhaysoos any chance to upgrade to the latest and see if this issue still persists?

- another suspect: are you doing like such `onClick={append}`? where the argument is not iterable
- are you append with a reference variable, which potentially gets mutated somehow?
- can you share a screenshot of your actual code

<img width="663" alt="Screen Shot 2022-01-10 at 3 20 24 pm" src="https://user-images.githubusercontent.com/10513364/148718117-32060be1-c635-4916-a3cd-dfd0100136d5.png">

```tsx
<input {...register(`contributions.${index}.value`)} />
```

you can omit that defaultValue="" as well

**Comment by dayhaysoos** [comment]
Omg @bluebill1049  it works after upgrading to latest! I hate myself for not checking this on my own holy shit!!

Thank you so much

**Comment by bluebill1049** [comment]
haha, awesome (sorry for the buggy old version)! no worries, I am glad it worked out.

**Comment by dayhaysoos** [comment]
I'm so sorry to open this backup! Before this error was happening when I used `append`. Now it only happens when I use `remove`.

In the code sandbox example, this error doesn't show for either.

To answer @bluebill1049 questions from earlier:

I'm not passing `onClick={append}`

I don't think I'm using a reference variable.

The actual code I have is in the code sandbox, unless there is more you'd like to see. Here's a loom video as well (I couldn't figure out how to end the recording immediately haha)

https://www.loom.com/share/35cc67007fc34e5cb965baff5cc12712

**Comment by bluebill1049** [comment]
@dayhaysoos can you share your entire code for the `useFieldArray` usage? If it's private you can send to me privately with twitter DM as well.

also include the entire error log as well (from chrome).

**Comment by dayhaysoos** [comment]
Hey @bluebill1049, I'm really confused because everything I am doing with useField Array is in the code sandbox I sent, but let me gather a bunch of gists for you:

This is where I'm using `useFieldArray`:

https://gist.github.com/dayhaysoos/c64ebef70480914a0bbdb991b1426c64


I also have this set up in a create page (I'm using Next.js).

WARNING!

The code here is super embarrassing haha, all you need to look at is like 215, where I'm giving the `<FormProvider />`, but I wanted to show you this just because it is a pretty complicated set up I have here and I might be missing something. If you have questions please let me know:

https://gist.github.com/dayhaysoos/535c863ba3e906043c5ff82f7fd0aac3


The log error:

```
    Uncaught TypeError: data is not iterable
    at removeAtIndexes (index.esm.mjs?184f:428)
    at removeArrayAt (index.esm.mjs?184f:437)
    at Object._updateFieldArray (index.esm.mjs?184f:1144)
    at remove (index.esm.mjs?184f:495)
    at onClick (contribution-input.js?a4f3:25)
    at HTMLUnknownElement.callCallback (react-dom.development.js?61bb:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js?61bb:3994)
    at invokeGuardedCallback (react-dom.development.js?61bb:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js?61bb:4070)
    at executeDispatch (react-dom.development.js?61bb:8243)
    at processDispatchQueueItemsInOrder (react-dom.development.js?61bb:8275)
    at processDispatchQueue (react-dom.development.js?61bb:8288)
    at dispatchEventsForPlugins (react-dom.development.js?61bb:8299)
    at eval (react-dom.development.js?61bb:8508)
    at batchedEventUpdates$1 (react-dom.development.js?61bb:22396)
    at batchedEventUpdates (react-dom.development.js?61bb:3745)
    at dispatchEventForPluginEventSystem (react-dom.development.js?61bb:8507)
    at attemptToDispatchEvent (react-dom.development.js?61bb:6005)
    at dispatchEvent (react-dom.development.js?61bb:5924)
    at unstable_runWithPriority (scheduler.development.js?3069:468)
    at runWithPriority$1 (react-dom.development.js?61bb:11276)
    at discreteUpdates$1 (react-dom.development.js?61bb:22413)
    at discreteUpdates (react-dom.development.js?61bb:3756)
    at dispatchDiscreteEvent (react-dom.development.js?61bb:5889)
```

Lemme know if you need to see anything else!

**Comment by bluebill1049** [comment]
> The code here is super embarrassing haha

hahaha, that's most of my code anyhow 

I will take a look at it during my lunch break. don't over stress on this 👍

**Comment by dayhaysoos** [comment]
I appreciate you! I've been trying really hard to solve on my own, wouldn't be here if I wasn't losing my shit haha. I will focus on other parts of my form in the mean time!

**Comment by bluebill1049** [comment]
hahaha i think i found out the issue!

**Comment by bluebill1049** [comment]
please remove those "bad boys" into it's own components: https://gist.github.com/dayhaysoos/535c863ba3e906043c5ff82f7fd0aac3#file-create-js-L143-L211

you are basically `unmount` and `re-mount` all those components on each re-render.

**Comment by dayhaysoos** [comment]
Oohh okay, so perhaps the form itself should be within the <Container /> component below, and everything that updates and re-renders should be inside the form. Is that what you meant? @bluebill1049 

Gonna start reworking it right now!

**Comment by bluebill1049** [comment]
```tsx
const Test = () => {
  const From = () => <form /> // don't do this
  
  return <Whatver />
}
```

**to**

```tsx
const From = () => <form /> // do this

const Test = () => {
  return <Whatver />
}
```

**Comment by dayhaysoos** [comment]
okay so I updated the gist with what I have now. Those two components now exist outside the original component, but still in the same file:

https://gist.github.com/dayhaysoos/535c863ba3e906043c5ff82f7fd0aac3

I'm still getting the `data is not iterable` error when I remove, only.

I just realized something, though! I am passing `useForm` in different instances of `<FormBuilder />`.

I think I should be using `useFormContext` there instead! I'll report back

**Comment by dayhaysoos** [comment]
False alarm, I was passing `useForm` before but I stopped using it and never removed the prop T_T

**Comment by bluebill1049** [comment]
see those guys still get mount and unmounted

https://gist.github.com/dayhaysoos/535c863ba3e906043c5ff82f7fd0aac3#file-create-js-L130-L170

**Comment by dayhaysoos** [comment]
Ahh, okay so what would be the best practice for a form where you're stepping through different fields? Vid for context:

https://www.loom.com/share/d860781ce9cf4f1ea2804d7c1eba2626

**Comment by bluebill1049** [comment]
- your first issue is component within component which result remount and unmount each re-render
- the second issue is here: 
```tsx
const stepMap = [
  {
    title: 'Personal Information',
    Component: (
      <FormBuilder
        formMap={personalInfoFormMap}
        form={useForm}
        candidateData={candidateData}
        title={'Personal Information'}
      />
    )
  },
  {
    title: 'Job Details',
    Component: (
      <FormBuilder
        formMap={jobDetailsFormMap}
        form={useForm}
        candidateData={candidateData}
        title={'Job Details'}
      />
    )
  },
  {
    title: 'Job History',
    Component: (
      <FormBuilder
        formMap={jobHistoryFormMap}
        form={useForm}
        candidateData={candidateData}
        title={'Job History'}
        skillTags={skillTags}
        setSkillTags={setSkillTags}
      />
    )
  },
  {
    title: 'Review',
    Component: <h1>Review Stuff</h1>
  }
]
```

This variable will get initialized on each render which leads to unmount and remount again and again as well.

## solutions

1. here is a solution of mine: https://react-hook-form.com/advanced-usage#WizardFormFunnel making each step a form 
2. Move the variable of components out from your root component, you can keep the variable but **not** hosting those components

**Comment by dayhaysoos** [comment]
Okay, thank you! I wish getting this working was as easy as removing the variable, but I think I might have to install React Router or something, this might take me some time but I really appreciate your help.

Will keep you updated on how it goes

**Comment by bluebill1049** [comment]
@dayhaysoos you don't need react-router, there is `push` in nextjs. Also, this variable issue can be resolved without removing it. 

```tsx
const stepMap = [
  {
    title: 'Personal Information',
  },
  {
    title: 'Job Details',
  },
  {
    title: 'Job History',
  },
  {
    title: 'Review',
  }
]

// turn it into a reusable component
function ReuseableComponent({ index }) {
  return <FormBuilder
    formMap={jobDetailsFormMap[index]}
    // form={useForm} context 
    candidateData={candidateData[index]}
    title={stepMap[index]}
  />
}
```

**Comment by dayhaysoos** [comment]
@bluebill1049 okay, I might be fundamentally not understanding something about React 😅 

So what I'm about to try is:

```
const formMap = [
  {
    title: 'Personal Information',
    formBuilderData: personalInfoFormMap
  },
  {
    title: 'Job Details',
    formBuilderData: jobDetailsFormMap
  },
  {
    title: 'Job History',
    formBuilderData: jobHistoryFormMap
  }
]
```

And then I have a FormSwitcher component (I think this is what you meant)


```
function FormSwitcher({ formMap, index }) {
  return (
    <FormBuilder
      formMap={formMap[index].formBuilderData}
      title={formMap[index].title}
    />
  )
}
```

This pretty much gives me the same results I've already had, the one thing I haven't incorporated is any kind of routing with React Router DOM or next/router. This is how I'm returning things:


```
  return (
    <Layout>
      <FormProvider {...methods}>
        <StepTracker
          formStep={formStep}
          setFormStep={setFormStep}
          stepMap={stepMap}
        />
        <Container sx={{ maxWidth: 476 }}>
          <FormSwitcher formMap={formMap} index={formStep} />
          <Stepper
            handleSubmit={handleSubmit}
            onSubmit={onSubmit}
            setFormStep={setFormStep}
            formStep={formStep}
            isDirty={isDirty}
          />
        </Container>
      </FormProvider>
    </Layout>
  )

```

Is this more or less what you meant with your advice? Sorry, if I'm misunderstanding something really easy.

**Comment by bluebill1049** [comment]
could you follow this codesandbox? https://codesandbox.io/s/distracted-sun-xm5ho?file=/src/App.js

```tsx
import React, { useState } from "react";
import {
  FormProvider,
  useFieldArray,
  useForm,
  useFormContext
} from "react-hook-form";
// i am only using this component because you do in your example
import { Stepper } from "react-form-stepper";

const Step1 = ({ onSubmit }) => {
  const { handleSubmit, register } = useFormContext();
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <input {...register("firstName")} placeholder="First Name" />
      <input type="submit" />
    </form>
  );
};

const Step2 = ({ onSubmit }) => {
  const { handleSubmit, register } = useFormContext();
  const { fields, append, remove } = useFieldArray({
    name: "test"
  });

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {fields.map((field, index) => {
        return (
          <React.Fragment key={field.id}>
            <input
              {...register(`test.${index}.name`)}
              placeholder="First Name"
            />
            <button type="button" onClick={() => remove(index)}>
              Remove
            </button>
          </React.Fragment>
        );
      })}
      <button type="button" onClick={() => append({ test: "data" })}>
        Append
      </button>
      <input type="submit" />
    </form>
  );
};

const Step3 = ({ onSubmit }) => {
  const { handleSubmit, register } = useFormContext();
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <input {...register("lastName")} placeholder="First Name" />
      <input type="submit" />
    </form>
  );
};

const Steps = ({ step, onSubmit }) => {
  switch (step) {
    case 0:
      return <Step1 onSubmit={onSubmit} />;
    case 1:
      return <Step2 onSubmit={onSubmit} />;
    case 2:
      return <Step3 onSubmit={onSubmit} />;
    default:
      return null;
  }
};

export default function App() {
  const [step, setStep] = useState(0);
  const methods = useForm();
  const onSubmit = (data) => {
    if (step === 2) {
      alert("You are done!");
    } else {
      setStep(step + 1);
    }
  };
  renderCount++;

  return (
    <FormProvider {...methods}>
      <Stepper
        steps={[{ label: "Step 1" }, { label: "Step 2" }, { label: "Step 3" }]}
        activeStep={step}
      />
      <Steps step={step} onSubmit={onSubmit} />
    </FormProvider>
  );
}
```

**Comment by dayhaysoos** [comment]
I'll try tomorrow, thanks, Bill!

**Comment by dayhaysoos** [comment]
Hey @bluebill1049, I made updates that are more similar to what you provided.

For some reason I'm still getting this error of `data is not iterable` when I try to use `remove()`. 

I played around with it, and despite seeing that error, it still seems to manage the data as expected. Everything seems to work fine in the code sandboxes and I can't figure out what's going on on my end. 

I think I might just close my eyes and pray nothing breaks in production at this point. If you're curious, here's what I have now:

https://gist.github.com/dayhaysoos/0e1c66a218b9964b71a52e90abe0b20e

If you have any more ideas on what might be the issue, I'm open to trying things. Otherwise I might just continue to deal with this on my own so I don't waste your time

**Comment by bluebill1049** [comment]
Hey, @dayhaysoos no worries about the investigation. I don't consider it as time waste 👍 This is really strange why it throws this error, I am considering swallow the error inside the library but it doesn't feel right to me why this is happening in the first place. Consider we have a huge list of tests both unit and automation and none of them caught this strange case. I will have a look at your gist again today, and keep this issue open for a few more days in case other people in the community caught a similar problem. Don't hesitate to ask more questions here as well.

**Comment by dayhaysoos** [comment]
okay!

I do have some really weird/complex things going on with my <FormBuilder />.. I think I'm going to get rid of the logic I'm using to automatically build forms and make them individual components. It's going to be a lot of work but perhaps somehow, somewhere , the issue will be addressed. Will report back. 

I agree, doesn't feel right to swallow the error, especially considering that all code sandbox instances work as expected.

**Comment by dayhaysoos** [comment]
Hey @bluebill1049 !!! I've got great news! With the refactor I'm doing, the error doesn't exist when I run `add` OR `remove`

I still don't fully understand what was wrong but it definitely had something to do with the way I set up my "form builder" component. It was only an issue for the dynamic form I was trying to make.

My new issue now is, for some reason, my input fields aren't applying the default value that I'm getting from initializing `useForm`. I just stumbled across this error though, so I will prob figure it out in due time.

**Comment by bluebill1049** [comment]
> Hey @bluebill1049 !!! I've got great news! With the refactor I'm doing, the error doesn't exist when I run add OR remove

nice! if your data is async, you should use `reset` at `useEffect` :) it may resolve your issue.

**Comment by dayhaysoos** [comment]
@bluebill1049  actually my error was there because I spelled something wrong :D 

I am so happy to get through this! This page itself is a little more complicated than basic forms, I need to add back file uploading (I already have this down) and add another dynamic form input. I think it's okay for me to close this, but I'ma leave it open until I'm done with this page, just in case.

Thank you so much for your help! This project has been amazing, when I launch this platform for https://blacktechpipeline.com in March, I plan on sponsoring this project!

**Comment by dayhaysoos** [comment]
Hey @bluebill1049, all seems to be going okay so far!

I have to do another similar dynamic input for collecting "skills". I want to gather skills like tags and allow candidates to submit that to the database.

I can't use useFieldArray again because I'm using it for "contributions" instead. I'm looking to use a third-party library to manage gathering these "skills" which I want in a flat array.

I was wondering if I could use `setValue` to store the skills array but I seem to have some trouble with getting react-hook-form to recognize this data. Do you have any advice on this kind of approach? Am I thinking about it incorrectly?

**Comment by bluebill1049** [comment]
if you don't need to shuffle or re-arrange data, you can just do 

add `setValue('test.0', 'data')`
remove `unregister('test.0')`

actually, if it's not related to form, you can just store your value in your state manamgent lib.

**Comment by dayhaysoos** [comment]
@bluebill1049  it is related to the form, it needs to be submitted with the rest of the form data to the database.

What I was doing before was just saving the array in useState and then submitting that whole array to the DB, but I read that react-hook-form doesn't support flat arrays. Going to go the setValue route, thank you

**Comment by bluebill1049** [comment]
IMHO you can still use useFieldArray tho, just do a data flatten at the submit before you hit the service.

**Comment by dayhaysoos** [comment]
hmm okay I'll consider this, right now I like that I have it working with a third party library right now. If this doesn't work out the way I hope, I'd go back to useFieldArray, but I might need help with that, though. Will report back! thank you!

**Comment by bluebill1049** [comment, marked as answer]
I am going to move this into a Github question since it's not an issue, feel free to leave question there 👍

**Reply by dayhaysoos** [reply]
Fine with me, I think I'm clear from the issues I was dealing with. I appreciate your help!

**Reply by bluebill1049** [reply]
👍  no worries

**Comment by dayhaysoos** [comment]
Hey @bluebill1049, having a hard time with something. To be as concise as possible:

Data coming from a backend is also used as defaultValue for the forms.

The data used to be `null` which made it so that if you typed in a field, and then deleted it, the state would still show as `isDirty` because the default value went from `null` to `""`.


I changed the default data from the API to be an empty string, however, I'm still seeing issues with typing in a field and deleting it. It still maintains `isDirty` state. Hopefully I make sense! Any idea what it might be?

**Reply by dayhaysoos** [reply]
False alarm! Seems to be working now but I'm not sure why 😅

**Reply by bluebill1049** [reply]
np, tip `isDirty` is pretty much doing a deepEqual with you defaultValues against your form values.
